src/lib/service/learnerRecordAPI/courseRecord/models/courseRecord.ts

```diff
@@ -0,0 +1,96 @@
+import { plainToClass } from 'class-transformer'
+
+import { Module } from '../../../../model'
+import { Record, RecordState } from '../../models/record'
+import { ModuleRecord } from '../../moduleRecord/models/moduleRecord'
+
+export enum CourseRecordPreference {
+	Liked = 'LIKED',
+	Disliked = 'DISLIKED',
+}
+
+export class CourseRecordResponse {
+	courseRecords: CourseRecord[]
+
+	constructor(courseRecords: CourseRecord[]) {
+		this.courseRecords = courseRecords
+	}
+}
+
+export class CourseRecord extends Record {
+	courseTitle: string
+	modules: ModuleRecord[]
+	preference?: CourseRecordPreference
+	lastUpdated?: Date
+	courseDisplayState?: string
+	required: boolean
+
+	constructor(
+		courseId: string,
+		userId: string,
+		state: RecordState = RecordState.Null,
+		modules: ModuleRecord[] = [],
+		courseTitle: string,
+		required: boolean,
+		preference?: CourseRecordPreference,
+		lastUpdated?: Date
+	) {
+		super(courseId, userId, state)
+		this.courseTitle = courseTitle
+		this.preference = preference
+		this.required = required
+		if (lastUpdated) {
+			this.lastUpdated = new Date(lastUpdated)
+		}
+
+		this.fillRecords(modules)
+	}
+
+	public upsertModuleRecord(moduleRecordId: number, moduleRecord: ModuleRecord) {
+		let existingModIndex
+		for (let i = 0; i < this.modules.length; i++) {
+			const mr = this.modules[i]
+			if (mr.id === moduleRecordId) {
+				existingModIndex = i
+			}
+		}
+
+		if (existingModIndex) {
+			this.modules[existingModIndex] = moduleRecord
+		} else {
+			this.modules.push(moduleRecord)
+		}
+	}
```

Comment:
This check to find if the module by ID exists could be shorthanded using the `find()` function.

```js
const moduleById = this.modules.find(module => module.id === moduleRecordId)

if(moduleById){
    // do stuff if module exists
}
else{
    // do stuff if module doesn't exist
}
```

Reply:
This isn't quite what I was after with this - however it has led me to find `findIndex`, which is far more efficient and nicer to read